File: src/main/java/programmers/게임맵최단거리.java
package programmers;

import java.util.*;

/**
 bfs 탐색
 */
class 게임맵최단거리 {
    int[] dr = new int[]{1, -1, 0, 0};
    int[] dc = new int[]{0, 0, 1, -1};
    int[][] visited;
    int n, m;
    public int solution(int[][] maps) {
        n = maps.length;
        m = maps[0].length;
        visited = new int[n][m];
        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[]{0, 0});
        visited[0][0] = 1;
        int cnt = 1;
        while(!queue.isEmpty()) {
            cnt++;
            int len = queue.size();
            for(int i = 0; i < len; i++) {
                int[] point = queue.poll();
                for(int d = 0; d < 4; d++) {
                    int r = point[0] + dr[d];
                    int c = point[1] + dc[d];

                    if(!isAvailable(maps, r, c)) {
                        continue;
                    }

                    if((r == n - 1) && (c == m - 1)) {
                        return cnt;
                    }

                    queue.offer(new int[]{r, c});
                    visited[r][c] = 1;
                }
            }
        }

        return -1;
    }

    private boolean isAvailable(int[][] maps, int r,int c) {
        if(r < 0 || n <= r || c < 0 || m <= c) {
            return false;
        }

        if(visited[r][c] == 1) {
            return false;
        }

        if(maps[r][c] == 0) {
            return false;
        }

        return true;
    }
}

File: 게임맵최단거리.cpp
#include<vector>
#include<queue>
#include<iostream>
using namespace std;

struct Node {
    int r;
    int c;
    int cnt;
};

int dr[] = {1, -1, 0, 0};
int dc[] = {0, 0, 1, -1};
bool visited[101][101];
queue<Node> q;

// bfs
int solution(vector<vector<int>> maps)
{
    int n = maps.size();
    int m = maps[0].size();
    
    Node node;
    node.r = 0;
    node.c = 0;
    node.cnt = 1;
    
    q.push(node);
    visited[0][0] = true;
    
    while(!q.empty()) {
        Node node = q.front();
        q.pop();
        
        if(node.r == n - 1 && node.c == m - 1) {
            return node.cnt;
        }
        
        for(int d = 0; d < 4; d++) {
            int nr = node.r + dr[d];
            int nc = node.c + dc[d];
            
            if(nr < 0 || nr >= n || nc < 0 || nc >= m) {
                continue;
            }
            
            if(visited[nr][nc]) {
                continue;
            }
            
            if(maps[nr][nc] == 0) {
                continue;
            }
            
            Node next;
            next.r = nr;
            next.c = nc;
            next.cnt = node.cnt + 1;
            
            q.push(next);
            visited[nr][nc] = true;
        }
    }
    
    return -1;
}
